buildscripts/generate_macros.py

```diff
@@ -19,77 +19,109 @@
 # This is used to to specify nested fields in the query builder.
 # BSON supports up to 100 levels of nesting.
 
+import StringIO
 import sys
 
 # variable for storing maximum BSON nesting depth, in case this changes.
 MAX_BSON_DEPTH = 100
 
+# A reference of what the generated macro should look like for n=3.
+# This is for use in testing and documentation.
+REFERENCE_MACRO_3 = (
+    "#define MONGO_ODM_CHILD3(base, field1, field2, field3) "
+    "make_nvp_with_parent(MONGO_ODM_KEY_BY_VALUE(&std::decay_t<typename decltype(MONGO_ODM_CHILD2("
+    "base, field1, field2))::no_opt_type>::field3), "
```

Comment:
Apologies for format changes & refactoring in the generate_macros.py script, but  the main change related to the bug can be seen on this line:
`MONGO_ODM_CHILD2(base, field1, field2))::type` became
`MONGO_ODM_CHILD2(base, field1, field2))::no_opt_type`